<a href="https://colab.research.google.com/github/jagtappranav2721-cpu/ML-Internship-FlyRank/blob/main/w02_ml_task_framing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-03 — Frame Your Lane as an ML Task

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane as an ML task (type)

*Classification, clustering, ranking, or scoring — which one, and why?*



The selected lane, **Refresh / Content Opportunity Scoring**, is primarily a **Ranking / Scoring** machine learning task. The objective is not simply to predict whether a content page should be refreshed, but to prioritize content pages based on their likelihood of benefiting from a review.

Rather than producing a binary yes/no decision, the expected output is a **priority score** that allows content teams to rank pages from highest to lowest review priority. This enables editors to focus their limited time and resources on the pages that are most likely to require attention.

Although a classification model may be used internally to estimate the probability that a page is declining, the final business objective is a ranked list of content pages. Therefore, the overall problem is best framed as a **Ranking / Scoring** task, where the output directly supports content prioritization and decision-making.

## 2. Target or proxy

*What would you predict? Where does that label come from — observed outcome or a defined rule?*



The ideal business objective is to identify content pages that are most likely to benefit from a refresh or review. However, the starter dataset does not contain a direct label indicating whether refreshing a page will improve its future performance.

Instead, I will use **`is_declining_label`** as the initial **proxy target**. This label indicates whether a content page is currently experiencing declining performance and serves as a reasonable indicator of pages that may require attention.

Using a proxy target allows the model to learn meaningful patterns from the available data while supporting the overall goal of prioritizing content for review. In a production environment, a stronger target could be developed using historical refresh actions and future performance data to directly measure the impact of content updates.

## 3. Success metric

*One metric you can defend. What number means 'good'?*


The primary objective of this project is to produce a ranked list of content pages that helps content teams prioritize their review efforts. Since editors typically review only the highest-priority pages, the evaluation metric should reflect the quality of those recommendations rather than overall prediction accuracy.

The primary success metric for this project is **Precision@K**, which measures how many of the top-ranked pages actually require review or refresh. A high Precision@K indicates that the model is successfully identifying the pages that are most valuable for editorial attention.

In addition to Precision@K, supporting metrics such as **Average Precision** and **Recall** can be used to evaluate the overall quality of the ranking and the model's ability to identify pages that genuinely require attention.

Overall, success is defined by the model's ability to prioritize the right pages for review, enabling content teams to make efficient and informed decisions with limited resources.

## 4. The unit of analysis, as a real dataframe

*Load your lane's slice and show it: one row = one what?*

In [12]:
!git clone https://github.com/jagtappranav2721-cpu/ML-Internship-FlyRank.git

fatal: destination path 'ML-Internship-FlyRank' already exists and is not an empty directory.


In [13]:
!find . -name "*.csv"

./ML-Internship-FlyRank/data/raw/content_refresh_anonymized.csv
./ML-Internship-FlyRank/outputs/refresh_queue_sample.csv
./sample_data/california_housing_test.csv
./sample_data/california_housing_train.csv
./sample_data/mnist_train_small.csv
./sample_data/mnist_test.csv


In [14]:
import pandas as pd

df = pd.read_csv("./ML-Internship-FlyRank/outputs/refresh_queue_sample.csv")
df.head()


,final_rank,content_id,client_id,final_refresh_score,best_model_name,best_model_probability,baseline_refresh_score,confidence,suggested_action,final_reason_codes,...,word_count,trend_direction,competition_level,content_type,main_intent,age_tier,freshness_tier,word_count_tier,impression_tier,position_tier
0,1,content_1f080331fa2b,client_3fdba35f04,81.636697,random_forest,0.782079,0.844481,high,refresh_and_review_ctr,declining_with_demand|low_ctr_visible_page|low...,...,1404.0,down,MEDIUM,keyword article,informational,91-180,91-180,1000-2000,good,page_1
1,2,content_6aa43079fb0c,client_3fdba35f04,81.447656,random_forest,0.788105,0.825477,high,refresh_and_review_ctr,declining_with_demand|low_ctr_visible_page|mod...,...,1457.0,down,LOW,keyword article,informational,91-180,91-180,1000-2000,good,page_1
2,3,content_d6570c51c9bd,client_3fdba35f04,81.430346,random_forest,0.847372,0.695884,medium,refresh_and_review_ctr,declining_with_demand|low_ctr_visible_page|mod...,...,1362.0,down,MEDIUM,keyword article,informational,91-180,91-180,1000-2000,moderate,striking
3,4,content_72e800a9c214,client_3fdba35f04,81.034960,random_forest,0.774371,0.842545,high,refresh_and_review_ctr,declining_with_demand|low_ctr_visible_page|mod...,...,1371.0,down,MEDIUM,keyword article,commercial,91-180,91-180,1000-2000,good,page_1
4,5,content_e04eb9549989,client_3fdba35f04,80.873188,random_forest,0.814805,0.749468,medium,refresh_and_review_ctr,declining_with_demand|low_ctr_visible_page|mod...,...,1408.0,down,LOW,keyword article,informational,91-180,91-180,1000-2000,good,page_1


In [15]:
lane_df = df[
    [
        "content_id",
        "impressions_90d",
        "clicks_90d",
        "sessions_90d",
        "avg_position",
        "ctr",
        "content_age_days",
        "days_since_last_update",
        "competition_level",
        "content_type",
        "main_intent",
        "is_declining_label"
    ]
]

print("Unit of Analysis:", "One row = One content page")
print("Number of rows:", len(lane_df))
print("Number of columns:", len(lane_df.columns))

lane_df.head()

Unit of Analysis: One row = One content page
Number of rows: 200
Number of columns: 12


,content_id,impressions_90d,clicks_90d,sessions_90d,avg_position,ctr,content_age_days,days_since_last_update,competition_level,content_type,main_intent,is_declining_label
0,content_1f080331fa2b,12834,6,66,6.8,0.05,165,104,MEDIUM,keyword article,informational,1
1,content_6aa43079fb0c,8064,6,23,3.8,0.07,139,104,LOW,keyword article,informational,1
2,content_d6570c51c9bd,2498,0,9,10.1,0.00,165,104,MEDIUM,keyword article,informational,1
3,content_72e800a9c214,13790,16,27,8.2,0.12,139,104,MEDIUM,keyword article,commercial,1
4,content_e04eb9549989,3393,3,5,3.6,0.09,131,104,LOW,keyword article,informational,1


In [18]:
lane_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 12 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   content_id              200 non-null    object 
 1   impressions_90d         200 non-null    int64  
 2   clicks_90d              200 non-null    int64  
 3   sessions_90d            200 non-null    int64  
 4   avg_position            200 non-null    float64
 5   ctr                     200 non-null    float64
 6   content_age_days        200 non-null    int64  
 7   days_since_last_update  200 non-null    int64  
 8   competition_level       200 non-null    object 
 9   content_type            200 non-null    object 
 10  main_intent             200 non-null    object 
 11  is_declining_label      200 non-null    int64  
dtypes: float64(2), int64(6), object(4)
memory usage: 18.9+ KB


In [16]:
lane_df[["content_id", "is_declining_label"]].head(10)

,content_id,is_declining_label
0,content_1f080331fa2b,1
1,content_6aa43079fb0c,1
2,content_d6570c51c9bd,1
3,content_72e800a9c214,1
4,content_e04eb9549989,1
5,content_b69288c5e701,1
6,content_9b6df29f7889,1
7,content_bb6ebb5ec8c8,1
8,content_4d76cdb3387b,1
9,content_b4f35d640b1c,1


In [17]:
lane_df["is_declining_label"].value_counts()

,count
is_declining_label,
1,197
0,3


###Unit of Analysis

The unit of analysis for this project is **one content page**. Each row in the dataset represents a single content page and contains information about its search performance, content characteristics, and refresh-related signals.

To better align the dataset with the selected lane, I displayed the columns that are most relevant for **Refresh / Content Opportunity Scoring**. These include metrics such as impressions, clicks, sessions, average search position, click-through rate (CTR), content age, days since the last update, competition level, content type, and search intent. Together, these features provide meaningful information for evaluating whether a content page may benefit from a refresh or review.

The current proxy target for this machine learning task is **`is_declining_label`**, which indicates whether a content page is currently experiencing declining performance. This column serves as the target that the model will learn to predict, while the remaining columns act as input features. Although `content_id` is included in the displayed dataframe to identify each page, it is an identifier rather than a predictive feature and would not be used during model training.

### Observation

The target variable (`is_declining_label`) is highly imbalanced in the starter dataset, with **197 pages labeled as declining** and **3 pages labeled as non-declining**. This observation reinforces the need for evaluation metrics that focus on identifying the most valuable pages for review, rather than relying solely on overall prediction accuracy.

## 5. Why ML beats a fixed rule here

*What makes the pattern too messy for an if-statement?*


A simple rule-based system could prioritize content pages using manually defined thresholds, such as refreshing pages with low CTR or content that has not been updated for a long time. While these rules may work for straightforward situations, they cannot effectively capture the complex relationships between multiple content and search performance signals.

In this project, the decision to prioritize a page depends on the combined influence of several factors, including impressions, clicks, sessions, average search position, content age, days since the last update, competition level, content type, and search intent. The importance of these factors may vary from one page to another, making it difficult to define a single set of rules that performs well in every situation.

A machine learning approach can learn these relationships directly from historical data and assign a priority score based on the interaction of multiple features. This enables the system to produce a more accurate and flexible ranking of content pages while still allowing content editors to make the final review and publishing decisions.

Therefore, machine learning provides a more scalable and reliable solution for content prioritization than relying solely on manually defined rules.

## Self-check

- [x] Every section above is filled — markdown thinking AND the code that backs it.
- [x] The notebook runs top to bottom with no errors (Runtime → Run all).
- [x] No client names, URLs, or private queries anywhere.
- [x] My claims use careful words: observed, measured, directional, decision-support.
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card.

### Reflection

This notebook frames the selected lane as a machine learning problem by identifying the task type, defining the proxy target, selecting appropriate evaluation metrics, and demonstrating why a machine learning approach provides better decision support than a fixed rule-based system.